improver/bases.py

```diff
@@ -0,0 +1,261 @@
+            raise ValueError(msg)
+        for time_coord_name in ["time", "forecast_reference_time"]:
+            time_coords = [c.coord(time_coord_name) for c in inputs]
+            if not all([tc == time_coords[0] for tc in time_coords[1:]]):
+                raise ValueError(
+                    f"{time_coord_name} coordinates do not match."
+                    "\n  "
+                    "\n  ".join(
+                        [str(c.name()) + ": " + str(c.coord("time")) for c in inputs]
+                    )
+                )
+
+    @staticmethod
+    def assert_spatial_coords_ok(spatial_matching_cubes: List[Cube]):
+        """
+        Raises appropriate ValueError if
+
+        - if the x and y coords are not exactly the same to the
+            precision of the floating-point values (this should be true for
+            any cubes derived using cube.regrid())
+
+        Can be overloaded where only a subset of inputs are expected to match.
+        Overloading functions can use self.get_cube()
+        """
+        if len(spatial_matching_cubes) <= 1:
+            return
+        if not spatial_coords_match(spatial_matching_cubes):
+            raise ValueError(
+                f"Spatial coords of input Cubes do not match: {spatial_matching_cubes}"
+            )
+
+    def parse_inputs(self, inputs: List[Cube], time_bounds=False) -> None:
+        """Extracts input cubes as described by self.cube_descriptors.
+
+        Args:
+            inputs:
+                List of Cubes containing exactly one of each input cube.
+            time_bounds:
+                True when all input cubes are expected to hate time bounds.
+        Raises:
+            ValueError:
+                If additional cubes are found
+        """
+        cubes = CubeList(inputs)
+        for desc in self.cube_descriptors.values():
+            desc._matched_name = desc.name
+            if desc.partial_name:
+                # Replace descriptor name with any cube name that contains the partial name
+                try:
+                    (desc._matched_name,) = [
+                        c.name() for c in cubes if desc.name in c.name()
+                    ]
+                except ValueError:
+                    pass  # This is picked up with a better error message later
+        expected_names = set(
+            [desc._matched_name for desc in self.cube_descriptors.values()]
+        )
+        cubes_names = set([cube.name() for cube in cubes])
+        diff = expected_names - cubes_names
+        if diff:
+            raise ValueError(f"Expected to find cube of {diff}, in {cubes_names}")
+        diff = cubes_names - expected_names
+        if diff:
+            raise ValueError(f"Unexpected Cube(s) found in inputs: {diff}")
+
+        spatial_matching_cubes = []
+        for attr, cube_values in self.cube_descriptors.items():
+            (cube,) = cubes.extract(cube_values._matched_name)
+            cube.convert_units(cube_values.units)
```

Comment:
Converting units is more a pre-processing step than a constraint handling.
I'm wondering where you might draw the line of what the plugin framework might do (i.e. at what point do we hand over this responsibility to the individual 'processing'.